src/libraries/System.Private.Xml/src/System/Xml/Serialization/ReflectionXmlSerializationReader.cs

```diff
@@ -1645,26 +1645,46 @@ private static XmlSerializationCollectionFixupCallback GetCreateCollectionOfObje
 
                     if (member.Source == null)
                     {
+                        var isList = mapping.TypeDesc.IsArrayLike && !mapping.TypeDesc.IsArray;
                         var pi = member.Mapping.MemberInfo as PropertyInfo;
-                        if (pi != null && typeof(IList).IsAssignableFrom(pi.PropertyType)
-                            && (pi.SetMethod == null || !pi.SetMethod.IsPublic))
+
+                        // Old serializers would trip over private property setters. Except for lists, which get special treatment.
+                        // So do the same here, even though this serializer is more capable of handling private setters. *sigh*
+                        // Do note that this still allows Lists with private setters to utilize that capability of a reflection serializer.
+                        if (!isList && pi != null && pi.SetMethod != null && !pi.SetMethod.IsPublic)
                         {
-                            member.Source = (value) =>
+                            member.Source = (value) => throw new InvalidOperationException(SR.Format(SR.XmlReadOnlyPropertyError, pi.Name, pi.DeclaringType!.FullName));
+                        }
+
+                        // If a list doesn't have a setter, or only has a private setter, we shouldn't just ignore it, because the default
+                        // constructor for the type might have created a list we can use.
+                        else if (isList && pi != null && (pi.SetMethod == null || !pi.SetMethod.IsPublic))
+                        {
+                            var addMethod = mapping.TypeDesc.Type!.GetMethod("Add");
+
+                            if (addMethod != null)
                             {
-                                var getOnlyList = (IList)pi.GetValue(o)!;
-                                if (value is IList valueList)
+                                member.Source = (value) =>
                                 {
-                                    foreach (var v in valueList)
+                                    var getOnlyList = pi.GetValue(o)!;
+                                    if (getOnlyList == null)
                                     {
-                                        getOnlyList.Add(v);
+                                        // No-setter lists should just be ignored if they weren't created by constructor. Private-setter lists are the exception.
+                                        if (pi.SetMethod != null && !pi.SetMethod.IsPublic)
+                                            throw new InvalidOperationException(SR.Format(SR.XmlReadOnlyPropertyError, pi.Name, pi.DeclaringType!.FullName));
```

Comment:
This if statement covers the case in the first outer if block on line 1654

Reply:
Right. That's the overall spirit of this. But the outer `if` on line 1654 applies to non-list properties. The reason we have to cover this case again for list properties is because the default constructor of our object might have created a list for us that we can use, and then we don't need a public setter. Which is why this code is here - it runs after the object is constructed and we check to see if the list property is null before complaining about a private setter. Again, it's goofy special case handling for lists that mimics the ILGen serializer.

Reply:
Comment structure updated.